rpm/0007-Add-RFC7636-aka-PKCE-support.patch

```diff
@@ -0,0 +1,127 @@
+very little harm if the server does not support it. However, if the
+service is broken and may not receive code_challenge, then there is
+SkipPKCE property that can be set to true.
+
+There is no support for "plain" method as the specification requires
+that implementations use "S256" when the device can handle sha256.
+Server must always support it anyway.
+
+Code challenge is generated just like the specification recommends,
+constructing 32-octet sequence that is then base64url encoded.
+
+Signed-off-by: Tomi Leppänen <tomi.leppanen@jolla.com>
+---
+ src/oauth2data.h     | 10 ++++++++++
+ src/oauth2plugin.cpp | 41 +++++++++++++++++++++++++++++++++++++++++
+ 2 files changed, 51 insertions(+)
+
+diff --git a/src/oauth2data.h b/src/oauth2data.h
+index 31184974f2cd25be7df5483415496a13f9df2cf0..7b4623b21a6678f7d32b4d4b1e8915191a058c29 100644
+--- a/src/oauth2data.h
++++ b/src/oauth2data.h
+@@ -125,6 +125,16 @@ namespace OAuth2PluginNS {
+          */
+         SIGNON_SESSION_DECLARE_PROPERTY(QString, TokenQuery);
+ 
++        /*!
++         * set to true to skip PKCE (RFC7636)
++         */
++        SIGNON_SESSION_DECLARE_PROPERTY(bool, SkipPKCE);
++
++        /*!
++         * code verifier for PKCE (RFC7636)
++         */
++        SIGNON_SESSION_DECLARE_PROPERTY(QByteArray, CodeVerifier);
++
+         /*!
+          * Any extra (non standard) parameters to send to the endpoint URI
+          */
+diff --git a/src/oauth2plugin.cpp b/src/oauth2plugin.cpp
+index a5f8e9769939ad1e0fe04840ef9208365167369b..aeeeb1afa8fc3cb05d7744c7b83d7bb35d5f6388 100644
+--- a/src/oauth2plugin.cpp
++++ b/src/oauth2plugin.cpp
+@@ -25,6 +25,8 @@
+ #include "oauth2plugin.h"
+ #include "oauth2tokendata.h"
+ 
++#include <random>
++#include <QCryptographicHash>
+ #include <QJsonDocument>
+ #include <QUrl>
+ #include <QUrlQuery>
+@@ -68,6 +70,9 @@ const QString CLIENT_ID = QString("client_id");
+ const QString CLIENT_SECRET = QString("client_secret");
+ const QString REFRESH_TOKEN = QString("refresh_token");
+ const QString AUTH_ERROR = QString("error");
++const QString CODE_CHALLENGE = QString("code_challenge");
++const QString CODE_CHALLENGE_METHOD = QString("code_challenge_method");
++const QString CODE_VERIFIER = QString("code_verifier");
+ 
+ const QByteArray CONTENT_TYPE = QByteArray("Content-Type");
+ const QByteArray CONTENT_APP_URLENCODED = QByteArray("application/x-www-form-urlencoded");
+@@ -75,6 +80,30 @@ const QByteArray CONTENT_APP_JSON = QByteArray("application/json");
+ const QByteArray CONTENT_TEXT_PLAIN = QByteArray("text/plain");
+ const QByteArray CONTENT_TEXT_HTML = QByteArray("text/html");
+ 
++namespace {
++
++QByteArray getRandomCodeVerifier()
++{
++    /* Random code verifier generation as PKCE (RFC7636) defines it */
++    std::random_device rd;
++    std::uniform_int_distribution<quint64> ud;
++    quint64 data[4] = { 0 };
++    for (unsigned int i = 0; i < 4; i++)
++        data[i] = ud(rd);
```

Comment:
I'm not exactly an expert in random number generation, but it at least used to be so that using std::random_device directly was not the de facto standard, as it may run out of entropy. Instead it's usually used for seeding e.g. [std::mt19937](http://en.cppreference.com/w/cpp/numeric/random/mersenne_twister_engine). But it might be that my info is outdated.

Reply:
> Perhaps [getrandom](https://man7.org/linux/man-pages/man2/getrandom.2.html) would be more like what I intended here and that would even allow simplifying the code a bit. Of course, that's very Linux specific and thus might be a little more difficult to upstream.

Uhm... Well, that would be nice if we used more modern Linux. Now I just get `ENOSYS`.

Reply:
Changed this to use /dev/urandom directly. At least that should guarantee our source to be random, but now it's much less portable than before. It's basically doing what `getrandom` would have done.

Reply:
> Uhm... Well, that would be nice if we used more modern Linux. Now I just get `ENOSYS`.
> 
> > getrandom() was introduced in version 3.17 of the Linux kernel.

We don't have any active port below that. Your issue is probably that linux-headers/uapi is to old.

Reply:
> > Uhm... Well, that would be nice if we used more modern Linux. Now I just get `ENOSYS`.
> > > getrandom() was introduced in version 3.17 of the Linux kernel.
> 
> We don't have any active port below that. Your issue is probably that linux-headers/uapi is to old.

I think you forgot that Jolla C is still at 3.10.

Additionally, our [kernel-headers](https://github.com/sailfishos/kernel-headers/releases/tag/3.18.136) package is at 3.18, which should be recent enough.